enhancements/ingress/global-options-enable-hsts.md

```diff
@@ -0,0 +1,411 @@
+
+last-updated: 2021-03-17
+
+status: implementable
+
+see-also:
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1512759
+  - https://bugzilla.redhat.com/show_bug.cgi?id=1430035
+
+replaces:
+
+superseded-by:
+
+---
+
+# Global Options to Enable HTTP Strict Transport Security (HSTS)
+
+This enhancement provides the capability to enable global automatic HSTS
+as a default for all routes.
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Operational readiness criteria is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In 3.x and 4.x customers can [provide a per-route annotation to enable HSTS](
+https://docs.openshift.com/container-platform/4.4/networking/routes/route-configuration.html#nw-enabling-hsts_route-configuration).  For
+customers with many routes or regulatory compliance issues, the manual per-route annotation is
+problematic.
+
+This enhancement extends the `IngressController` API to allow cluster administrators to
+enable HSTS globally, without having to add an annotation to each route.
+
+## Motivation
+
+HSTS ([RFC 6797](https://tools.ietf.org/html/rfc6797)) policy enforces the use
+of HTTPS in client requests to the host, without having to use an HTTP redirect.
+HSTS provides user protection and is concerned with minimizing security threats based on network
+traffic eavesdropping and man-in-the-middle attacks.  Using a response header
+called `Strict-Transport-Security`, the HTTP response informs clients that a website
+can only be accessed via HTTPS.  
+
+Administrators who are tasked with route management and/or regulatory compliance
+issues want an option to automatically and globally enable HSTS. They also want
+to separately apply HSTS globally to worker nodes and infrastructure nodes (e.g.
+Prometheus, Grafana, web console hosts), while retaining the ability to opt-out
+of HSTS on a per-route basis.
+
+### Goals
+
+This proposal allows administrators to:
+- Continue to enable HSTS per-route (retain the existing feature functionality)
+- Enable HSTS for all routes globally, based on the cluster's ingress domain
+- Enable HSTS per-router
+- Disable HSTS per-route when it is enabled globally
+
+### Non-Goals
+
+- HSTS cannot be applied to non-TLS routes, even if HSTS is requested for all routes globally.  As
+mentioned in [RFC 6797 Section 8.1](https://tools.ietf.org/html/rfc6797#section-8.1), if an HTTP
+response is received over insecure transport, clients must ignore any present STS headers.
+- The current implementation has a `preload` directive for the `Strict-Transport-Security`
+header.  This is not an RFC 6797 directive and therefore its implementation may vary by
+browser.  It can be specified, but support may vary by implementation.
```

Comment:
Is this a non-goal?  Looks like it belongs under "Risks and Mitigations" or "Implementation Details/Notes/Constraints".